### ISSUE TITLE
Depend on js-beautify directly

### ISSUE THREAD

**Issue by octref** [octref/vetur#82]
`vscode-html-languageservice` depends on `js-beautify` directly. However, it's not aware of scss, less regions. It's also causing problems like #77.

`js-beautify` actually supports scss and less, so vetur will just depend on it directly.

Overall:

- html/css/scss/less/js formatting will be supported using js-beautify
- Have a default setting for each supported languages, but also expose settings passing to js-beautify, for example:
  - `"vetur.vue-html.format.config": ".htmlbeautifyrc"`
  - `"vetur.css.format.config": ".cssbeautifyrc"`
  - `"vetur.js.format.config": ".jsbeautifyrc"`
- Other languages don't have any reliable formatters yet, so just disable them all
- If someone is interested in having formatting for a specific language, open a new issue with link to an actively maintained formatter

**Comment by octref** [comment]
Will use https://github.com/mrmlnc/vscode-pugbeautify for jade/pug.

**Comment by beeplin** [comment]
wonderful~! waiting for the new version~~

**Comment by octref** [comment]
Formatters are temporarily disabled until this feature is addressed.

**Comment by zema1** [comment]
Looking forward to the new version..
I really need the formatter..

**Comment by zengyiqun** [comment]
I really need the formatter too !~~~~

**Comment by sebastianmacias** [comment]
+1

**Comment by christianmalek** [comment]
@octref Couldn't you just disable the formatter per default and make it accessible per workspace settings? I don't think that the added new lines are such an big problem.

**Comment by rafaelpimpa** [comment]
It was working fine for me for basic setups, wasn't it? Would like to see it enabled back.

**Comment by rafaelpimpa** [comment]
It was working fine for basic setups, wasn't it? Would like to see it enabled back.

**Comment by octref** [comment]
This is top on my list. Give me a day or two to work on it.

**Comment by octref** [comment]
So here is how it goes. Notice html/css/js/scss/less are all formatted, whereas sass/stylus/pug are left alone.

![format](https://cloud.githubusercontent.com/assets/4033249/23886720/7f1feadc-0853-11e7-9189-f73d78be2eab.gif)

Will add initial indentation to the template part and remove the initial indentation on js and make a release.

**Comment by octref** [comment]
Stole these config from https://github.com/victorporof/Sublime-HTMLPrettify

```ts
const htmlOptions = {
  brace_style: 'collapse', // [collapse|expand|end-expand|none] Put braces on the same line as control statements (default), or put braces on own line (Allman / ANSI style), or just put end braces on own line, or attempt to keep them where they are
  end_with_newline: false, // End output with newline
  indent_char: ' ', // Indentation character
  indent_handlebars: false, // e.g. {{#foo}}, {{/foo}}
  indent_inner_html: false, // Indent <head> and <body> sections
  indent_scripts: 'keep', // [keep|separate|normal]
  indent_size: 2, // Indentation size
  max_preserve_newlines: 0, // Maximum number of line breaks to be preserved in one chunk (0 disables)
  preserve_newlines: true, // Whether existing line breaks before elements should be preserved (only works before elements, not inside tags or for text)
  unformatted: ['a', 'span', 'img', 'code', 'pre', 'sub', 'sup', 'em', 'strong', 'b', 'i', 'u', 'strike', 'big', 'small', 'pre', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'], // List of tags that should not be reformatted
  wrap_line_length: 0 // Lines should wrap at next opportunity after this number of characters (0 disables)
};

const cssOptions = {
  end_with_newline: false, // End output with newline
  indent_char: ' ', // Indentation character
  indent_size: 2, // Indentation size
  newline_between_rules: true, // Add a new line after every css rule
  selector_separator: ' ',
  selector_separator_newline: true // Separate selectors with newline or not (e.g. 'a,\nbr' or 'a, br')
};
```

These will be used as initial default, let me know what you agree & don't agree with.
I'll add capability to specify your own js-beautify config later.

**Comment by octref** [comment]
Check out 0.5.2 and let me know how it works.

Still yet to be done:

- [ ] allow to specify own options
- [ ] pug formatter

**Comment by beeplin** [comment]
pressing shift-alt-f for multiple times will increase the indentation of html `<!-- -->` comments:
![image](https://cloud.githubusercontent.com/assets/13058150/24007147/b52de942-0aa8-11e7-9ac5-93c0d86c0e62.png)
became
![image](https://cloud.githubusercontent.com/assets/13058150/24007164/c320077e-0aa8-11e7-9c42-7d798c787569.png)
and became
![image](https://cloud.githubusercontent.com/assets/13058150/24007179/cdc08ed8-0aa8-11e7-92a9-ef00ca0b03af.png)

**Comment by octref** [comment]
0.5.4 preserves new lines for html, and uses `force-aligned` to align attributes as @beeplin suggested in #77 

Before:
![image](https://cloud.githubusercontent.com/assets/4033249/24007255/65805880-0a44-11e7-83c3-09986f246d13.png)

After:
![image](https://cloud.githubusercontent.com/assets/4033249/24007286/7fc3ebf8-0a44-11e7-9c4d-eae39b09f098.png)

Seems css/scss/less preserve new lines option is being added this weekend, https://github.com/beautify-web/js-beautify/pull/1146

I'll add it once it lands in `js-beautify`.

**Comment by beeplin** [comment]
v0.5.4 is wonderful :)

**Comment by octref** [comment]
0.5.6 brings CSS `preserve_newlines` from js-beautify, and is set as default:

![format2](https://cloud.githubusercontent.com/assets/4033249/24106009/14e70aba-0d5d-11e7-8963-6c2d5e537571.gif)

**Comment by danikane** [comment]
Thanks to @beeplin for submitting the bug, but please, please, please @octref find a way to specify:
	"html.format.wrapAttributes": "auto"
as you mentioned previously in this thread (allow to specify own options).
Thank you for your work!

**Comment by octref** [comment]
@danikane Will do soon.

**Comment by octref** [comment]
@danikane This is already in master so closing. Also `auto` is set as default but people can set `wrap_attributes` to `force-aligned` if they want. Will be available in 0.6.

**Comment by danikane** [comment]
Many thanks!

**Comment by x8x** [comment]
@octref Any news on status of pugbeautify integration?

**Comment by octref** [comment]
@x8x I forgot that. You can track https://github.com/vuejs/vetur/issues/527.

**Comment by x8x** [comment]
@octref Amazing! Thx! :-)
